### ISSUE TITLE
Updating parent entity's updatedAt field when updating/creating/deleting child entity in one-to-many or one-to-one relationship

### ISSUE THREAD

**Issue by akashkashyap** [prisma/prisma#17274]
## Problem

If a solution to this already exists and we are missing something, please point us to it!

Currently there is no way to automatically trigger an update to the parent entity's updatedAt field when a child entity is created/updated/deleted. If you want to do this, you have to manually write code in every location to handle updating the parent entity's updatedAt field. This is extremely cumbersome and prone to bugs.

## Suggested solution

There should be a configurable option to have updates to child fields affect parents' updatedAt timestamps. This should be configurable from the parent side, so that parent entities can choose how they want to handle this. E.g. 2 parents may share a relationship to the same children and each parent may want to handle updatedAt differently for the same child.

## Alternatives

An alternative would be having a generic api for side effects in Prisma. E.g., being able to add the equivalent of configurable callbacks that trigger arbitrary code when a certain set of changes happen on a given field. This is essentially a more generic version of the suggested solution.

## Additional context

There have been other discussions on this issue but I didn't see anything addressed by the Prisma team. This is a major issue for us in our application right now. I considered adding this as a bug, because it feels like a tablestakes feature that is missing. We'd love to either see this addressed and/or understand the reasoning if this issue was explicitly not addressed.

Other discussions:

1. https://github.com/prisma/prisma/discussions/10420
2. https://stackoverflow.com/questions/70806367/prisma-how-to-automatically-update-updatedat-field-of-parent-element-when-a-ch

**Comment by janpio** [comment]
Would it also be a valid solution to introduce an explicit attribute similar to `@updatedAt`, e.g. `@childrenUpdatedAt`, that is triggered when one of the explicit children of a model is updated? It could have a level argument `@childrenUpdatedAt(3)` to configure the depth.

I am asking because the suggested solution "configurable option to have updates to child fields affect parents' updatedAt timestamps" easily can lead to _really_ long chain of updates - technically even circles. So we need some limit to stop that - and even though that could be a global option as well (and default to a sane value), it led me to think that essentially this configuration belongs to the model and its time field that will be updated.

The side effect API could probably be implemented with https://www.prisma.io/blog/client-extensions-preview-8t3w27xkrxxn - but that would live in code and not in the schema, so has the danger of getting out of sync more easily.

How do other ORMs and database clients implement this feature?

**Comment by akashkashyap** [comment]
Your suggestion could work, but I think it would be simpler and more readable to explicitly set the option on the child instead of the parent. The ergonomics seem preferable.

As an example of how this is solved in other ORM's, ActiveRecord let's you specify "touch: true" on the belongs_to relation: [](https://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-touch).

I haven't looked into the internals of how exactly it's implemented in Rails, but at the product feature level I would expect this to work the same way in Prisma. If you only allow setting this from the child definition, then you solve the problem of unintentional long chains because a child can only update it's parent via the touch. Users can explicitly set touch up the parent chain as far as they want to. Loops can be addressed specifically for self relations, etc.

Here is an example of how I would expect this to work with a Prisma schema where a User can have many members. The @touch option used on the user relation within the Member model would do what we want:

`
model User {
  id            Int      @id @default(autoincrement())
  createdAt     DateTime @default(now())
  updatedAt     DateTime @updatedAt
  members       Member[]
  name          String?
  email         String
}

model Member {
  id                      Int             @id @default(autoincrement())
  createdAt               DateTime        @default(now())
  updatedAt               DateTime        @updatedAt
  user                    User            @relation(fields: [userId], references: [id], onDelete: Cascade) **@touch**
  userId                  Int
`

**Comment by Kirkify** [comment]
Ya, Laravel Eloquent follows a similar implementation, check that out here, would love to have something similar with Prisma :)
https://laravel.com/docs/10.x/eloquent-relationships#touching-parent-timestamps

**Comment by codinginflow** [comment]
I'm waiting for the same feature. Would the new [client extensions](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/query#modify-a-specific-operation-in-a-specific-model) API be a good way to implement this manually meanwhile?

**Comment by gurpreetatwal** [comment]
Related: https://github.com/prisma/prisma/issues/12492

**Comment by edelbalso** [comment]
I want to put in a +1 for this and #12492. It unblocks some really simple cacheing and data replication patterns for our APIs.

**Comment by Red-Killer** [comment]
+1

**Comment by vasilii-kovalev** [comment]
Would be nice to have it for many-to-many relationship as well.
I've added some comment in the mentioned discussion: https://github.com/prisma/prisma/discussions/10420#discussioncomment-11577848.

**Comment by anandakelvin** [comment]
```
prisma.Parent.update({
  where: {
    id: parent.d,
  },
  data: {
    Child: {
      update: {
        where: {
          id: child.id,
        },
        data,
      },
    },
  },
})
```

This way it "virtually" updates the parent, thus touching `@updatedAt` of parent

**Comment by akashkashyap** [comment]
Crazy this still hasn't been addressed, I get notifications from people posting on this thread every few months. For anyone reading this, we gave up on prisms a while ago and switched to drizzle once it became stable.

**Comment by dizzyjaguar** [comment]
any updates on this?

**Comment by AdhamMoussa** [comment]
any updates on this?
